### ISSUE TITLE
Timing-aware ISS model

### ISSUE THREAD

**Issue by chili-chips-ba** [chili-chips-ba/wireguard-fpga#3]
With ISS as one abstraction level option in Wireguard-FPGA [sim TB](https://github.com/chili-chips-ba/wireguard-fpga?tab=readme-ov-file#simulation-test-bench), we are looking for it to be timing-aware. And so in a structured, plug-and-play way. 

This mean that, while it does not need to be fully cycle-accurate, but only "close enough", the ISS should also allow easy swaps and user-accessible adjustments of its timing model. This [article](https://www.linkedin.com/posts/simon-southwell-7684482_riscv-iss-embeddedsoftware-activity-7256311551178027008-1nlZ/?utm_source=share&utm_medium=member_desktop) presents what is currently available. What seems to be missing is:
- ability to specify clock frequency
- standardized method for customizing timing parameters to any given RISC-V CPU, possibly with a worked out library of a few popular options*
- timing model of caches, possibly with following two options:
> - Statistical -- Random spread from 0 to a given n, where n can be set by user
> - Functional -- Configurable for size of lines, number of ways and number of sets
____
(*)
1) **Ibex** [timing model](https://ibex-core.readthedocs.io/en/latest/03_reference/pipeline_details.html#id1).

2) **PicoRV32** timing model should use the following settings:
- 5 cycles for Load and Store
- 4 cycles for everything else
That's for on-chip TCM as it currently does not have any Caching option.

3) **eduBOS5** [info](https://www.canva.com/design/DAGMkqkJ7_8/eEyKWi66g7BRhRAOZif7WA/view?utm_content=DAGMkqkJ7_8&utm_campaign=designshare&utm_medium=link&utm_source=editor) card shows that in its standard "2-stage" config, the core consumes:
- 3 cycles for Load and Store
- 2 cycles for everything else

eduBOS5 "3-stage" pipeline option adds 1 cycle to everything. It in return allows faster clock.

**Comment by wyvernSemi** [comment]
Why is there a need to specify clock frequency when using the ISS for simulation? All the units for timing are in cycles, which is what the ISS/co-sim synchronisation uses. Whatever clock speed it (or a logic implementation) is run at, it will still take that many cycles. The frequency only affects the generation of the clock in the test bench (and the synthesis constraints, of course).

The ISS already has a really simple method to update seven defined categories of timing. Maybe an external class can be defined with a set of constants, one set for each supported implementation, and a method to update the ISS, with the timing set selected at construction.

The cache timing (as in my LatticMico32 ISS) is actually external to the ISS, and done in the number of reported wait states returned by the external memory access callback function. So this can be made to be one of a cached or non-cached wait state values. The current WireGuard ISS integration code does not have a cache model, but the LatticeMico32 code should be adaptable.

**Comment by chili-chips-ba** [comment]
> **Why is there a need to specify clock frequency when using the ISS for simulation?** 

You are right, there isn't. It's simply the matter of getting used to the fact that the ISS is in this case merely a [VProc](https://github.com/wyvernSemi/vproc) app. And, the VProc has an input for bringing in whatever clock the actual CPU is running at. 

Traditionally, the Instruction Set Simulators have no notion of time, or clock. But, we are parting away with traditions 😉.
![image](https://github.com/user-attachments/assets/140502df-849b-48d9-a934-db58cae19f78)

---------------
> **an external class can be defined with a set of constants, one set for each supported implementation, and a method to update the ISS, with the timing set selected at construction.**

Yes, that would do 👍 

---------------
> **cache timing (as in my LatticMico32 ISS) is actually external to the ISS**

Since [Ibex](https://github.com/lowRISC/ibex), our front-runner RISC-V core includes Caches as an internal config option, we are good as the Cache model is contained within VProc
![image](https://github.com/user-attachments/assets/4d736858-bac4-412f-a07f-d58b9a08b9dd)

**Comment by chili-chips-ba** [comment]
Update from @wyvernSemi:
> "... ISS now includes means to select a timing model from a small selection of predefined cores (which can be added to):
>- rv32_timing_config::risc_v_core_e::DEFAULT,
>- rv32_timing_config::risc_v_core_e::PICORV32
>- rv32_timing_config::risc_v_core_e::EDUBOS5STG2
>- rv32_timing_config::risc_v_core_e::EDUBOS5STG3
>- rv32_timing_config::risc_v_core_e::IBEXMULSGL
>- rv32_timing_config::risc_v_core_e::IBEXMULFAST
>- rv32_timing_config::risc_v_core_e::IBEXMULSLOW
>
> There is also means to `enable and configure instruction cache timings` where the wait states are added depending on hit or miss. The cache can be configured (within limits) for the:
>- size of a cache line
>- number of ways
>- number of sets
>- memory region that is cached. 
>
> The penalty for a single non-cached word access can also be specified, separate from the normal load/store timing (which is assumed to be cached). That allows calculation of filling a cache line from slower memory. Please refer to [README](https://github.com/chili-chips-ba/wireguard-fpga?tab=readme-ov-file#risc-v-compiled-application) for additional detail.

The next step is for a DV junior on Wireguard team to pick this up, reproduce, and start using. We are therefore leaving this discussion track open for follow up questions...

**Comment by wyvernSemi** [comment]
The rv32 ISS integration code now allows selection of the timing model from the `vusermain.cfg` ISS configuration file (the [README](https://github.com/chili-chips-ba/wireguard-fpga?tab=readme-ov-file#risc-v-compiled-application) is updated with details).

There is now also some [documentation](https://github.com/chili-chips-ba/wireguard-fpga/tree/main/4.sim/models/rv32#rv32-iss-model-integration-for-wireguard) in the 4.sim/models/rv32 directory with some internal details of the VProc code that integrates the rv32 ISS, including how and where to add new timing models.

As far as I'm aware, there are no further outstanding actions on this issue, but not closing as per the last comment.
